### PR TITLE
Allow package installation and usage with fallback fonts

### LINKED ISSUE #39: Installation fails on Ubuntu AWS EC2 instance
Installation fails during testing of loading from temporary location. Stdout readout below:

```
Downloading GitHub repo Arcadia-Science/arcadiathemeR@HEAD
These packages have more recent versions available.
It is recommended to update all of them.
Which would you like to update?

Installing 1 packages: glue
trying URL 'https://cloud.r-project.org/src/contrib/glue_1.8.0.tar.gz'
Content type 'application/x-gzip' length 126682 bytes (123 KB)
==================================================
downloaded 123 KB

* installing *source* package ‘glue’ ...
** package ‘glue’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
using C compiler: ‘x86_64-conda-linux-gnu-cc (conda-forge gcc 14.1.0-0) 14.1.0’
x86_64-conda-linux-gnu-cc -I"/home/ubuntu/miniconda3/envs/popgenR/lib/R/include" -DNDEBUG   -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/ubuntu/miniconda3/envs/popgenR/include -I/home/ubuntu/miniconda3/envs/popgenR/include -Wl,-rpath-link,/home/ubuntu/miniconda3/envs/popgenR/lib   -fvisibility=hidden -fpic  -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/ubuntu/miniconda3/envs/popgenR/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/r-base-split_1714471492496/work=/usr/local/src/conda/r-base-4.3.3 -fdebug-prefix-map=/home/ubuntu/miniconda3/envs/popgenR=/usr/local/src/conda-prefix  -c glue.c -o glue.o
x86_64-conda-linux-gnu-cc -I"/home/ubuntu/miniconda3/envs/popgenR/lib/R/include" -DNDEBUG   -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/ubuntu/miniconda3/envs/popgenR/include -I/home/ubuntu/miniconda3/envs/popgenR/include -Wl,-rpath-link,/home/ubuntu/miniconda3/envs/popgenR/lib   -fvisibility=hidden -fpic  -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/ubuntu/miniconda3/envs/popgenR/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/r-base-split_1714471492496/work=/usr/local/src/conda/r-base-4.3.3 -fdebug-prefix-map=/home/ubuntu/miniconda3/envs/popgenR=/usr/local/src/conda-prefix  -c init.c -o init.o
x86_64-conda-linux-gnu-cc -I"/home/ubuntu/miniconda3/envs/popgenR/lib/R/include" -DNDEBUG   -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/ubuntu/miniconda3/envs/popgenR/include -I/home/ubuntu/miniconda3/envs/popgenR/include -Wl,-rpath-link,/home/ubuntu/miniconda3/envs/popgenR/lib   -fvisibility=hidden -fpic  -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/ubuntu/miniconda3/envs/popgenR/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/r-base-split_1714471492496/work=/usr/local/src/conda/r-base-4.3.3 -fdebug-prefix-map=/home/ubuntu/miniconda3/envs/popgenR=/usr/local/src/conda-prefix  -c trim.c -o trim.o
x86_64-conda-linux-gnu-cc -shared -L/home/ubuntu/miniconda3/envs/popgenR/lib/R/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,--allow-shlib-undefined -Wl,-rpath,/home/ubuntu/miniconda3/envs/popgenR/lib -Wl,-rpath-link,/home/ubuntu/miniconda3/envs/popgenR/lib -L/home/ubuntu/miniconda3/envs/popgenR/lib -o glue.so glue.o init.o trim.o -L/home/ubuntu/miniconda3/envs/popgenR/lib/R/lib -lR
installing to /home/ubuntu/miniconda3/envs/popgenR/lib/R/library/00LOCK-glue/00new/glue/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (glue)

The downloaded source packages are in
        ‘/tmp/RtmpycMfJa/downloaded_packages’
Updating HTML index of packages in '.Library'
Making 'packages.html' ... done
Running `R CMD build`...
* checking for file ‘/tmp/RtmpycMfJa/remotes1028d76a0f818/Arcadia-Science-arcadiathemeR-cd6def7/DESCRIPTION’ ... OK
* preparing ‘arcadiathemeR’:
* checking DESCRIPTION meta-information ... OK
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
Omitted ‘LazyData’ from DESCRIPTION
* building ‘arcadiathemeR_0.1.0.tar.gz’
* installing *source* package ‘arcadiathemeR’ ...
** using staged installation
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘arcadiathemeR’:
 .onAttach failed in attachNamespace() for 'arcadiathemeR', details:
  call: data.frame(fontfile = ttfiles, FontName = "", stringsAsFactors = FALSE)
  error: arguments imply differing number of rows: 0, 1
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/home/ubuntu/miniconda3/envs/popgenR/lib/R/library/arcadiathemeR’
Warning message:
In i.p(...) :
  installation of package ‘/tmp/RtmpycMfJa/file1028d16b41e29/arcadiathemeR_0.1.0.tar.gz’ had non-zero exit status
  ```

### PR DESCRIPTION
**Pull request by wasimsandhu**
## Summary
Fixes #39, in which the following error was thrown on package installation:
```
Error: package or namespace load failed for ‘arcadiathemeR’:
 .onAttach failed in attachNamespace() for 'arcadiathemeR', details:
  call: data.frame(fontfile = ttfiles, FontName = "", stringsAsFactors = FALSE)
  error: arguments imply differing number of rows: 0, 1
```

This error was occurring because `load_fonts.R` was trying to import "Suisse" as a font regardless of whether it existed or not.
```R
font_import(pattern = "Suisse", prompt = FALSE)
```

This PR imports the system default fallback font(s) if "Suisse" is not available, which fixes the installation error. However, upon usage, I discovered that the Suisse fonts had been hardcoded in, so I had to dynamically set those as well. I'm not familiar with R so please let me know if there's a more optimal approach.

## Checklist

- [x] Fix installation error due to import of unavailable custom font
- [x] Use fallback sans font in plots if custom font is unavailable
- [x] Test install and usage on macOS and Linux platforms

### CODE REVIEW

**Comment by wasimsandhu** [comment]
Looks like arcadiathemeR was not meant to be used without the Suisse font, which makes sense given that these plots are meant to adhere to Arcadia style guides. This is what they look like without the Suisse font:

Screenshot 2024-11-05 at 10 39 42 AM

Tested installation on macOS and Ubuntu, both working now.

**Comment by keithchev** [comment]
> Looks like arcadiathemeR was not meant to be used without the Suisse font, which makes sense given that these plots are meant to adhere to Arcadia style guides. This is what they look like without the Suisse font:

Do you mean that the plots just look weird/wrong or that there are additional errors or warnings? As long as the package works without errors, I think it's okay, because we can assume users will know (or figure out) to install the Suisse fonts for full compliance w the style guide.

**Comment by wasimsandhu** [comment]
@keithchev No, I mean that I had to make changes in this diff so that the package could be used without the Suisse font (see PR description, which I updated a little earlier today). But yes, I agree that the theme is usable without the font.

File: R/load_fonts.R [hunk 1]
```diff
@@ -8,24 +7,34 @@
...
+  available_fonts <- extrafont::fonts()
+  font_family <- custom_font
+  if (!(custom_font %in% available_fonts)) {
+    font_family <- system(sprintf("fc-match -f '%%{family}' %s", fallback_font), intern = TRUE)
+  }
+
+  # Ubuntu returns "DejaVu Sans" while R looks for "DejaVuSans", so remove spaces
+  if (Sys.info()["sysname"] == "Linux") {
+    font_family <- gsub(" ", "", font_family)
+  }
```
[the remark this answers is not in the document]
Comment (wasimsandhu) [reply R/load_fonts.R:20]: Ah yes you're right

**Comment by wasimsandhu** [comment]
@keithchev Hey, yeah I tested installation and it appeared that the fonts were loaded correctly, but I did not test usage. There was a flaw in my conditional check (was checking if pattern "Suisse" appeared in the list of available fonts instead of an actual font such as "Suisse Int'l"). 

TLDR: Re-tested installation and usage on macOS with and without Suisse fonts, both cases appear to be working as expected. I will retest on Linux after Taylor gets a chance to review

File: R/arcadia.R [hunk 2]
```diff
@@ -85,6 +86,16 @@ SEMIBOLD_FONT <- "Suisse Int'l Semi Bold"
 MEDIUM_FONT <- "Suisse Int'l Medium"
 MONO_FONT <- "Suisse Int'l Mono"
 
+# Use OS default sans fallback if Suisse is not installed
+FONT <- NULL
```
Comment (taylorreiter) [inline R/arcadia.R:90]: what does `FONT <- NULL` protect against here? will it auto default to sans or something?

File: R/load_fonts.R [hunk 3]
```diff
@@ -8,24 +7,34 @@
 #' @export
 #'
 load_arcadia_fonts <- function(custom_font = "Suisse", fallback_font = "sans") {
-  # define font names to check for
-  font_names <- c("Suisse Int'l", "Suisse Int'l Semi Bold", "Suisse Int'l Medium", "Suisse Int'l Mono")
+  # check if custom font available
+  font_family <- custom_font
+
+  if (!("Suisse Int'l" %in% extrafont::fonts())) {
```
Comment (taylorreiter) [inline R/load_fonts.R:13]: is hardcoding here a problem? is there a way to use `custom_font` here instead? (since you removed hardcoding of `font_names`

File: R/load_fonts.R [hunk 4]
```diff
@@ -8,24 +7,34 @@
...
+
   if (length(missing_fonts) == 0) {
     cat(sprintf("All custom fonts '%s' are successfully loaded.\n", paste(font_names, collapse = ", ")))
     return(font_names)
   } else {
     cat(sprintf("Custom fonts '%s' not found. Using fallback font '%s'.\n", paste(missing_fonts, collapse = ", "), fallback_font))
     return(rep(fallback_font, length(font_names)))
   }
-}
+}
```
Comment (taylorreiter) [inline R/load_fonts.R:40]: nit: missing new line EOF

**Review by taylorreiter — APPROVED** [review]
This looks reasonable to me but a few small comments/questions! 

Given the back and forth with Keith earlier in this review, it may also be worth opening an issue to automate this testing in the future!

File: R/load_fonts.R (hunk 3 above)
Comment (wasimsandhu) [reply R/load_fonts.R:13]: Font names are hardcoded, so I think it's okay to keep it this way for now... If font names change, we would have to update this script anyway.

File: R/load_fonts.R [hunk 5]
```diff
@@ -8,24 +7,34 @@
 #' @export
 #'
 load_arcadia_fonts <- function(custom_font = "Suisse", fallback_font = "sans") {
-  # define font names to check for
-  font_names <- c("Suisse Int'l", "Suisse Int'l Semi Bold", "Suisse Int'l Medium", "Suisse Int'l Mono")
+  # check if custom font available
+  font_family <- custom_font
```
[the remark this answers is not in the document]
Comment (wasimsandhu) [reply R/load_fonts.R:11]: Yeah you're right, the renaming is confusing, probably because the comments are out of sync with the code now. Let me refactor this so it's clearer that we are defining the font to import as either `custom_font` or `fallback_font`.

File: R/arcadia.R (hunk 2 above)
Comment (wasimsandhu) [reply R/arcadia.R:90]: Copilot mistake I forgot to remove 😅

File: R/load_fonts.R (hunk 4 above)
Comment (wasimsandhu) [reply R/load_fonts.R:40]: Fixed. Is there a good R formatter for vscode you can recommend?

**Comment by wasimsandhu** [comment]
Refactored font import into a try-catch statement. I think this is more concise and readable. Re-tested installation and usage on macOS and Linux, everything looks good with and without the Suisse font!

@taylorreiter Agreed on opening an issue for automated testing. The testing matrix of macOS/Linux + with/without custom font has been super tedious to get through.

File: R/load_fonts.R (hunk 4 above)
Comment (taylorreiter) [reply R/load_fonts.R]: I don't think I know one unfortunately :( lintr is the "standard" R linter, but it isn't as superb as some of the python ones
